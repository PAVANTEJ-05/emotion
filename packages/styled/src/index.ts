import React from 'react'
import { Theme } from '@emotion/react'
import styled from './base'
import { tags } from './tags'
import {
  CreateStyledComponent,
  CreateStyled as BaseCreateStyled
} from './types'
export type {
  ArrayInterpolation,
  ComponentSelector,
  CSSObject,
  FunctionInterpolation,
  Interpolation
} from '@emotion/serialize'
export type {
  CreateStyledComponent,
  FilteringStyledOptions,
  StyledComponent,
  StyledOptions
} from './types'

export type StyledTags = {
  [Tag in keyof React.JSX.IntrinsicElements]: CreateStyledComponent<
    {
      theme?: Theme
      as?: React.ElementType
    },
    React.JSX.IntrinsicElements[Tag]
  >
}

export interface CreateStyled extends BaseCreateStyled, StyledTags {}

// bind it to avoid mutating the original function
const newStyled = styled.bind(null) as CreateStyled

tags.forEach(tagName => {
  ;(newStyled as any)[tagName] = newStyled(tagName as keyof typeof newStyled)
})

export default newStyled
