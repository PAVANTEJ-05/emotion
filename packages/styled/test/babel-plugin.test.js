import 'test-utils/setup-env'
import React from 'react'
import { act } from 'react'
import { render } from '@testing-library/react'
import styled from '@emotion/styled'

test("config merging works even if it's referenced by variable", async () => {
  const Button = ({ isRed, ...rest }) => (
    <button {...rest}>{isRed ? 'forwarded' : 'not forwarded'}</button>
  )

  const cfg = { shouldForwardProp: p => p !== 'isRed' }
  const StyledButton = styled(Button, cfg)({})
  const { container } = render(<StyledButton isRed />)
  expect(container).toMatchSnapshot()
})
