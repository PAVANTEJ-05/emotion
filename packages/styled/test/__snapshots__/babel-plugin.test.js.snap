// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config merging works even if it's referenced by variable 1`] = `
<div>
  <button
    class="emotion-0 emotion-1"
  >
    not forwarded
  </button>
</div>
`;
